pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -221,18 +256,30 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
     assert _partialUpsertHandler != null;
     AtomicReference<GenericRow> previousRecordReference = new AtomicReference<>();
+    AtomicBoolean outOfOrder = new AtomicBoolean();
     RecordLocation currentRecordLocation = _primaryKeyToRecordLocationMap.computeIfPresent(
         HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), (pk, recordLocation) -> {
           if (recordInfo.getComparisonValue().compareTo(recordLocation.getComparisonValue()) >= 0) {
-            _reuse.clear();
-            previousRecordReference.set(recordLocation.getSegment().getRecord(recordLocation.getDocId(), _reuse));
+            IndexSegment currentSegment = recordLocation.getSegment();
```

Comment:
Add an if check of `if (!recordInfo.isDeleteRecord())`. No need to read previous record when current record is delete